File: algorithmAndDataStructure/golang/longestIncreasingSubsequence/longestIncreasingSubsequence.go
package longestIncreasingSubsequence

func lower_bound(nums []int, value int) (x int) {
	y := len(nums)
	for x < y {
		m := x + (y-x)/2
		if nums[m] >= value {
			y = m
		} else {
			x = m + 1
		}
	}
	return
}

func lengthOfLIS(nums []int) int {
	numsLen := len(nums)
	lowEndings := make([]int, 0, numsLen)
	for _, num := range nums {
		i := lower_bound(lowEndings, num)
		lowEndingsLen := len(lowEndings)
		if i >= lowEndingsLen {
			lowEndings = append(lowEndings, num)
		} else {
			lowEndings[i] = num
		}
	}
	return len(lowEndings)
}

File: algorithmAndDataStructure/cpp/longestIncreasingSubsequence/longestIncreasingSubsequence.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int maxValue = 0, numsLen = nums.size();
        vector<int> dp(nums);
        for (int i = 0; i < numsLen; i++) {
            dp[i] = 1;
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[i], dp[j]+1);
                }
            }
            maxValue = max(maxValue, dp[i]);
        }
        return maxValue;
    }

    int lengthOfLISII(vector<int>& nums) {
        int numsLen = nums.size();
        vector<int> lowEndings;
        for (int i = 0; i < numsLen; i++) {
            auto it = lower_bound(lowEndings.begin(), lowEndings.end(), nums[i]);
            if (it == lowEndings.end()) {
                lowEndings.push_back(nums[i]);
            } else {
                *it = nums[i];
            }
        }
        return lowEndings.size();
    }
};